pkg/controller/gkenetworkparamset/gnpcontroller_validations.go

```diff
@@ -0,0 +1,246 @@
+			isParamsNewer := params.CreationTimestamp.After(otherGNP.CreationTimestamp.Time)
+
+			if isDifferentGNP && isMatchingVPC && isParamsNewer {
+				return &gnpValidation{
+					IsValid:      false,
+					ErrorReason:  networkv1.DeviceModeVPCAlreadyInUse,
+					ErrorMessage: fmt.Sprintf("GNP with deviceMode can't reference a VPC already in use. VPC '%s' is already in use by '%s'", otherGNP.Spec.VPC, otherGNP.Name),
+				}, nil
+			}
+
+			if isDifferentGNP && isMatchingSubnet && isParamsNewer {
+				return &gnpValidation{
+					IsValid:      false,
+					ErrorReason:  networkv1.DeviceModeSubnetAlreadyInUse,
+					ErrorMessage: fmt.Sprintf("GNP with deviceMode can't reference a subnet already in use. Subnet '%s' is already in use by '%s'", otherGNP.Spec.VPC, otherGNP.Name),
+				}, nil
+			}
+		}
+	}
+
+	//if GNP with deviceMode and The referencing VPC is the default VPC
+	if params.Spec.DeviceMode != "" {
+		networkResource, err := cloud.ParseResourceURL(c.gceCloud.NetworkURL())
+		if err != nil {
+			return nil, err
+		}
+		if params.Spec.VPC == networkResource.Key.Name {
+			return &gnpValidation{
+				IsValid:      false,
+				ErrorReason:  networkv1.DeviceModeCantUseDefaultVPC,
+				ErrorMessage: "GNP with deviceMode can't reference the default VPC",
+			}, nil
+		}
+	}
+
+	return &gnpValidation{IsValid: true}, nil
+}
+
+type gnpNetworkCrossValidation struct {
+	IsValid      bool
+	ErrorReason  networkv1.GNPNetworkParamsReadyConditionReason
+	ErrorMessage string
+}
+
+func (val *gnpNetworkCrossValidation) toCondition() v1.Condition {
+	condition := v1.Condition{}
+
+	if val.IsValid {
+		condition.Status = v1.ConditionTrue
+	} else {
+		condition.Status = v1.ConditionFalse
+		condition.Reason = string(val.ErrorReason)
+		condition.Message = val.ErrorMessage
+	}
+
+	condition.Type = string(networkv1.NetworkConditionStatusParamsReady)
+	condition.LastTransitionTime = v1.NewTime(time.Now())
+
+	return condition
+}
+
+func crossValidateNetworkAndGnp(network *networkv1.Network, params *networkv1.GKENetworkParamSet) *gnpNetworkCrossValidation {
+	if network.Spec.Type == networkv1.L3NetworkType {
+		if params.Spec.PodIPv4Ranges == nil {
+			return &gnpNetworkCrossValidation{
+				IsValid:      false,
+				ErrorReason:  networkv1.L3SecondaryMissing,
+				ErrorMessage: "L3 type network requires secondary range to be specified in params",
+			}
+		}
+		if params.Spec.DeviceMode != "" {
+			return &gnpNetworkCrossValidation{
+				IsValid:      false,
+				ErrorReason:  networkv1.L3DeviceModeExists,
+				ErrorMessage: "L3 type network can't be used with a device mode specified in params",
+			}
+		}
+	}
+
+	if network.Spec.Type == networkv1.DeviceNetworkType {
+		if params.Spec.DeviceMode == "" {
+			return &gnpNetworkCrossValidation{
+				IsValid:      false,
+				ErrorReason:  networkv1.DeviceModeMissing,
+				ErrorMessage: "Device type network requires device mode to be specified in params",
+			}
+		}
+		if params.Spec.PodIPv4Ranges != nil {
```

Comment:
Maybe we should check for empty as well as nil?

Reply:
good catch.